### PR TITLE
feat(api): workspace and clip CRUD API (US-9.4)

### LINKED ISSUE #78: US-9.4: Workspace and clip CRUD API
## User Story

**As a** musician, **I want to** create, list, update, and delete workspaces and clips via the API, **so that** I can organize my music library from any client.

## Description

Expose the workspace and clip management operations as REST endpoints. Clips support search, filtering, and sorting. A default workspace is created when a new user is registered.

## Functional Requirements

- Workspace endpoints: `POST /api/v1/workspaces`, `GET /api/v1/workspaces`, `GET /api/v1/workspaces/{id}`, `PATCH /api/v1/workspaces/{id}`, `DELETE /api/v1/workspaces/{id}`
- Clip endpoints: `GET /api/v1/clips` (list with pagination), `GET /api/v1/clips/{id}`, `PATCH /api/v1/clips/{id}`, `DELETE /api/v1/clips/{id}`
- Clip list supports: `?workspace_id=`, `?search=`, `?style=`, `?bpm_min=`, `?bpm_max=`, `?key=`, `?model=`, `?sort=newest|oldest`, `?page=`, `?per_page=`
- Deleting a workspace with clips requires `?force=true` or returns 409
- All endpoints scoped to the authenticated user

## Acceptance Criteria

- [ ] Full CRUD lifecycle works for workspaces (create, list, get, update, delete)
- [ ] Full CRUD lifecycle works for clips (list, get, update, delete)
- [ ] Clip search filters by style, BPM range, key, and model
- [ ] Pagination returns correct page counts and respects per_page limits
- [ ] Deleting a non-empty workspace without `?force=true` returns 409

## Dependencies

**Blocked by:** Stage 8 (auth + workspaces/clips collections — #70, #71). Wraps CLI workspace/clip logic from Stage 4.

## References

- User stories: `user-stories/02-layer-2-platform-api.md` — Stage 9, US-9.4

### PR DESCRIPTION
**Pull request by frankbria**
## Summary
Implements #78: US-9.4 Workspace and clip CRUD API

- **Workspace endpoints** (`/api/v1/workspaces`): POST (201; 409 on duplicate name per user, backed by a new unique `(user_id, name)` index), GET list with per-workspace clip counts (single aggregation), GET/PATCH/DELETE by id. DELETE returns 409 for a non-empty workspace without `?force=true`, 400 for the user's last workspace; a forced delete cascades clips and their stored audio.
- **Clip CRUD endpoints** (`/api/v1/clips`): GET list with filters (`workspace_id`, `search`, `style`, `bpm_min`/`bpm_max`, `key`, `model`), `newest|oldest` sort, and pagination (`page`, `per_page` default 20 / max 100, `total_pages`); GET/PATCH/DELETE by id (PATCH updates title only; DELETE removes the stored audio object first). Search/style needles are regex-escaped, case-insensitive substrings — mirrors the CLI's `search_clips`.
- **Default workspace at registration**: the OAuth callback bootstraps every account via the idempotent `get_or_create_default_workspace` (moved from the generation service to the new workspaces service; generation keeps a lazy fallback). A pre-existing non-default workspace named "My Workspace" is promoted instead of colliding with the unique index.
- **Stale-token hardening**: new `require_existing_user` dependency — CRUD endpoints resolve the token's user first, so a valid token for a deleted account gets 404 instead of creating orphaned records (mirrors `POST /generate`).

## Acceptance Criteria
- [x] Full CRUD lifecycle works for workspaces (create, list, get, update, delete)
- [x] Full CRUD lifecycle works for clips (list, get, update, delete)
- [x] Clip search filters by style, BPM range, key, and model
- [x] Pagination returns correct page counts and respects per_page limits
- [x] Deleting a non-empty workspace without `?force=true` returns 409
- [x] All endpoints scoped to the authenticated user; default workspace created at registration

## Test Plan
- [x] Tests written first (TDD): `tests/test_workspaces_api.py` (38), `tests/test_clips_crud_api.py` (37), auth-callback bootstrap tests
- [x] All tests passing: 1125 unit + 127 API integration (real MongoDB, real JWT, no mocking)
- [x] Diff coverage 98% on changed lines (gate ≥85%); the 10 uncovered lines are concurrency race-recovery branches
- [x] Linting clean (black + ruff)
- [x] Internal code review (advisory) completed — 1 Major fixed (bootstrap name-collision promotion)
- [x] Cross-family review pass: codex — 2 P2 findings fixed (stale-token 404)
- [x] Test mutation sanity check completed — 6 mutations (force/409 check, ownership check, pagination skip, regex escape, default promotion, stale-user 404), all caught

## Known Limitations / Intentionally Deferred
- **Clip CRUD is strictly owner-scoped**: another user's clip returns 404 even when `is_public=true`. Public visibility applies only to the US-9.3 audio endpoint — listing/browsing public clips is a future discovery feature.
- **`ClipResponse` omits `file_path`**: storage keys are internal; clients fetch audio via `GET /clips/{id}/audio`. (Deviation from the CodeRabbit plan, which listed it.)
- **Forced workspace delete removes stored audio sequentially** (one object per clip). Fine at current scale; revisit with bounded concurrency if workspaces grow to thousands of clips.
- **Unique `(user_id, name)` index on workspaces**: any pre-existing deployment with duplicate workspace names per user would need a dedupe before this index builds. No production deployment exists yet, so no migration is shipped.
- **Workspace-delete preconditions are checked, not transactional**: the last-workspace and clip-count guards are pre-checks; a concurrent delete could race them. Atomic enforcement needs multi-document transactions (replica-set-only) or a per-user lock — neither exists in this deployment. Worst case (user transiently at zero workspaces) is self-healing via the lazy default-workspace bootstrap. (CodeRabbit finding, rebutted in review thread.)
- **API test fixtures are duplicated per test module** (matches the pre-existing pattern across all six API test files); consolidating into `tests/conftest.py` is a follow-up refactor beyond this PR's scope.
- **Clip `updated_at`** is not tracked (the Clip model has no such field); adding one is out of scope for this issue.
- **Date-range clip filters** (`date_from`/`date_to` from the CLI) are not exposed — the issue's query-param list does not include them.

## Implementation Notes
Deviations from the CodeRabbit plan on the issue (its research predates Stages 8–9): schemas live in router modules per codebase convention (no `schemas/` package); services follow the existing plural naming (`services/workspaces.py`, extended `services/clips.py`); the existing `routers/clips.py` (US-9.3 audio) was extended rather than recreated; tests follow the established integration pattern (httpx.AsyncClient + isolated throwaway Mongo DB + real JWTs) instead of a mocked `get_current_user` override; `workspace_id` is optional on `GET /clips` since clips carry `user_id` directly.

Closes #78

## Summary by CodeRabbit

* **New Features**
  * Workspace management: create, list, fetch, rename, and delete workspaces with clip counts and idempotent default-workspace bootstrap
  * Full clips CRUD: owner-scoped list, get, rename (title-only), and delete clips; audio streaming preserved

* **Behavior Changes**
  * Auth tightened: tokens for deleted users now yield 404
  * Workspace names unique per user; delete enforces constraints (force cascade)

* **Search & Filtering**
  * Advanced clip search: title/style substring, BPM ranges, key, model, sorting, pagination

* **Tests**
  * Extensive integration tests for workspaces, clips, auth edge cases, and default-workspace bootstrapping

* **Documentation**
  * API docs updated for Workspaces and expanded Clips endpoints

### CODE REVIEW

File: src/acemusic/api/routers/clips.py [hunk 1]
```diff
@@ -1,33 +1,196 @@
...
+    page: int = Field(default=1, ge=1)
+    per_page: int = Field(default=PER_PAGE_DEFAULT, ge=1, le=PER_PAGE_MAX)
+
+    @model_validator(mode="after")
+    def _check_bpm_range(self) -> "ClipSearchParams":
+        # An inverted range can never match; reject it as the client error it
+        # is instead of silently returning an empty page.
+        if self.bpm_min is not None and self.bpm_max is not None and self.bpm_min > self.bpm_max:
+            raise ValueError("bpm_min must not be greater than bpm_max.")
+        return self
```
[the remark this answers is not in the document]
Comment (frankbria) [reply src/acemusic/api/routers/clips.py:75]: Fixed in c43c5eb — blank `search`/`style` values are normalized to `None` by a field validator, with a regression test (`test_blank_search_and_style_filters_are_ignored`).

File: src/acemusic/api/routers/clips.py [hunk 2]
```diff
@@ -1,33 +1,196 @@
...
+    clip_id: str,
+    body: ClipUpdate,
+    current: CurrentUser = Depends(require_existing_user),
+) -> ClipResponse:
+    """Rename the clip; an empty body is a no-op returning the current state."""
+    if body.title is None:
+        clip = await clip_service.get_owned_clip(clip_id, current.user_id)
+    else:
+        clip = await clip_service.update_clip_title(clip_id, current.user_id, body.title)
+    return ClipResponse.from_clip(clip)
```
[the remark this answers is not in the document]
Comment (frankbria) [reply src/acemusic/api/routers/clips.py:202]: Fixed in c43c5eb — an explicit `"title": null` now returns 422 via a model validator using `model_fields_set`; an omitted field remains a documented no-op. Regression test added.

File: src/acemusic/api/services/workspaces.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+    if clip_count:
+        storage = get_storage_backend()
+        async for clip in Clip.find(Eq(Clip.workspace_id, workspace.id)):
+            # delete() does file/network I/O via the sync backend; keep it off
+            # the event loop.
+            await asyncio.to_thread(storage.delete, clip.file_path)
+        await Clip.find(Eq(Clip.workspace_id, workspace.id)).delete()
+
+    await workspace.delete()
```
[the remark this answers is not in the document]
Comment (frankbria) [reply src/acemusic/api/services/workspaces.py:184]: Acknowledged but intentionally deferred: enforcing these preconditions atomically requires multi-document transactions (replica-set-only — local dev and CI run standalone mongod) or a per-user lock collection, neither of which exists in this deployment. The worst-case race outcome (a user transiently at zero workspaces) is benign and self-healing: `get_or_create_default_workspace` lazily recreates the default on the next login or generation job. Recording this in the PR's Known Limitations; happy to revisit when the deployment moves to a replica set.

File: src/acemusic/api/routers/clips.py [hunk 4]
```diff
@@ -1,33 +1,213 @@
...
+    per_page: int = Field(default=PER_PAGE_DEFAULT, ge=1, le=PER_PAGE_MAX)
+
+    @field_validator("search", "style")
+    @classmethod
+    def _blank_to_none(cls, value: str | None) -> str | None:
+        # `?search=` must mean "no filter"; an empty needle would otherwise
+        # become an empty regex that matches every clip with the field set.
+        if value is None:
+            return None
+        return value.strip() or None
```
[the remark this answers is not in the document]
Comment (frankbria) [reply src/acemusic/api/routers/clips.py:67]: Keeping this comment: it records a non-obvious WHY (an empty needle silently becomes a match-all regex — the bug this validator prevents), which the validator name alone doesn't convey. This matches the codebase's convention of annotating constraint rationale (see the surrounding routers/services).

File: src/acemusic/api/routers/clips.py [hunk 5]
```diff
@@ -1,33 +1,213 @@
...
+            raise ValueError("Clip title must not be blank.")
+        return value
+
+    @model_validator(mode="after")
+    def _reject_explicit_null(self) -> "ClipUpdate":
+        # An omitted title is a no-op; an explicit null is a malformed request
+        # (title has no "cleared" state) and must not masquerade as success.
+        if "title" in self.model_fields_set and self.title is None:
+            raise ValueError("title must be a non-empty string; omit the field to leave it unchanged.")
+        return self
```
[the remark this answers is not in the document]
Comment (frankbria) [reply src/acemusic/api/routers/clips.py:101]: Keeping this comment for the same reason: it documents the contract distinction (omitted = no-op, explicit null = malformed) that the code's model_fields_set check doesn't make self-evident. Consistent with the repo's annotation style.

**Comment by frankbria** [comment]
## Final Triage Summary (PR #118)

Cutoff: 2026-06-11T02:00:11Z (last push, 16c4175)
New findings since cutoff: 2 CodeRabbit Minor + 4 claude-review observations (all explicitly non-blocking). No Critical/Major.

### Fixed (previous round, commit c43c5eb)
- [coderabbitai] Blank `?search=`/`?style=` became match-all regex → normalized to no-filter, regression test added
- [coderabbitai] Explicit `"title": null` indistinguishable from omitted → now 422 via `model_fields_set`, test added
- [claude-review] `WorkspaceCreate` validator wiring → documented `@field_validator` + `@classmethod` pattern
- [claude-review] Test counter → `itertools.count`

### Skipped (with justification)
- [coderabbitai, Major, prev. round] Non-atomic workspace-delete preconditions — rebutted in thread: atomic enforcement needs replica-set transactions or a lock collection (absent in this deployment); worst case is benign and self-healing via the lazy default-workspace bootstrap. Recorded in Known Limitations.
- [coderabbitai, Minor ×2] Remove validator comments — kept: they document non-obvious WHY (match-all-regex hazard; omitted-vs-null contract), matching the repo's annotation convention.
- [claude-review] `total_pages=0` on empty result — mathematically consistent; no frontend consumer exists yet to dictate the `max(1, …)` convention (YAGNI).
- [claude-review] Combined `style`+`search` filter test — AND semantics are structural (independent keys in one Mongo query dict); each filter is independently tested.
- [claude-review] `_coerce_object_id` duplicated in two services — extraction deferred until a third user appears.
- [claude-review] `require_existing_user` per-request user lookup — acknowledged trade-off; correctness over a negligible round-trip at current scale.
- [claude-review, prev. round] Test fixture consolidation into conftest — touches all six API test modules; follow-up refactor beyond this PR's scope.
- [claude-review, prev. round] GET/PATCH workspace `clip_count` soft race — transient off-by-one; the list endpoint is authoritative via aggregation.
- [internal review] Sequential storage deletes on force-delete — in Known Limitations; no scale requirement today.

CI: all checks green on 16c4175 (ci 3.11/3.12, CodeRabbit, GitGuardian, claude-review).
Demo: all acceptance criteria verified with outcome evidence (Phase 11).
